crates/executor/src/blockchain_tree/mod.rs

```diff
@@ -0,0 +1,811 @@
+    fn insert_chain(&mut self, chain: Chain) -> ChainId {
+        let chain_id = self.chain_id_generator;
+        self.chain_id_generator += 1;
+        self.block_indices.insert_chain(chain_id, &chain);
+        // add chain_id -> chain index
+        self.chains.insert(chain_id, chain);
+        chain_id
+    }
+
+    /// Insert block inside tree. recover transaction signers and
+    /// internaly call [`BlockchainTree::insert_block_with_senders`] fn.
+    pub fn insert_block(&mut self, block: SealedBlock) -> Result<bool, Error> {
+        let senders = block.senders().ok_or(ExecError::SenderRecoveryError)?;
+        let block = SealedBlockWithSenders::new(block, senders).unwrap();
+        self.insert_block_with_senders(&block)
+    }
+
+    /// Insert block with senders inside tree
+    pub fn insert_block_with_senders(
+        &mut self,
+        block: &SealedBlockWithSenders,
+    ) -> Result<bool, Error> {
+        // check if block number is inside pending block slide
+        let last_finalized_block = self.block_indices.last_finalized_block();
+        if block.number <= last_finalized_block {
+            return Err(ExecError::PendingBlockIsFinalized {
+                block_number: block.number,
+                block_hash: block.hash(),
+                last_finalized: last_finalized_block,
+            }
+            .into())
+        }
+
+        // we will not even try to insert blocks that are too far in future.
+        if block.number > last_finalized_block + self.num_of_side_chain_max_size {
+            return Err(ExecError::PendingBlockIsInFuture {
+                block_number: block.number,
+                block_hash: block.hash(),
+                last_finalized: last_finalized_block,
+            }
+            .into())
+        }
+
+        // check if block is already inside Tree
+        if self.block_indices.contains_block_hash(block.hash()) {
+            // block is known return that is inserted
+            return Ok(true)
+        }
+
+        // check if block is part of canonical chain
+        if self.block_indices.canonical_hash(&block.number) == Some(block.hash()) {
+            // block is part of canonical chain
+            return Ok(true)
+        }
+
+        // check if block parent can be found in Tree
+        if let Some(parent_chain) = self.block_indices.get_blocks_chain_id(&block.parent_hash) {
+            self.fork_side_chain(block.clone(), parent_chain)?;
+            //self.db.tx_mut()?.put::<tables::PendingBlocks>(block.hash(), block.unseal())?;
+            return Ok(true)
+        }
+
+        // if not found, check if it can be found inside canonical chain.
+        if Some(block.parent_hash) == self.block_indices.canonical_hash(&(block.number - 1)) {
+            // create new chain that points to that block
+            self.fork_canonical_chain(block.clone())?;
+            //self.db.tx_mut()?.put::<tables::PendingBlocks>(block.hash(), block.unseal())?;
```

Comment:
Remove?

Reply:
Added comments and issue: https://github.com/paradigmxyz/reth/issues/1713